include/bluetooth/audio/cap.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_CAP_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_CAP_H_
+
+/**
+ * @brief Common Audio Profile (CAP)
+ *
+ * @defgroup bt_cap Common Audio Profile (CAP)
+ *
+ * @ingroup bluetooth
+ * @{
+ *
+ * [Experimental] Users should note that the APIs can change
+ * as a part of ongoing development.
+ */
+
+#include <zephyr/types.h>
+#include <bluetooth/audio/csis.h>
+#include <bluetooth/audio/tbs.h>
+#include <bluetooth/audio/media_proxy.h>
+#include <bluetooth/audio/audio.h.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Type of CCID records */
+enum bt_cap_ccid_record_type {
+	/** CCID belongs to a Telephone Bearer Service instance */
+	BT_CAP_CCID_TYPE_TELEPHONE_BEARER,
+	/** CCID belongs to a Media Control Service instance */
+	BT_CAP_CCID_TYPE_MEDIA_PLAYER
+};
+
+/** CCID record structure */
+struct bt_cap_ccid_record {
+	/** The type of the record */
+	enum bt_cap_ccid_record_type type;
+
+	union {
+		/** Reference to a local or remote Telephone Bearer instance */
+		struct bt_tbs *telephone_bearer;
+		/** Reference to a local or remote Media Control instance */
+		struct media_player *media_player;
+	};
+};
+
+/** Callback structure for CAP procedures */
+struct bt_cap_cb {
+	/**
+	 * @brief CCID find callback
+	 *
+	 * Callback for bt_cap_find_ccid_service().
+	 * This is called for the first service instance found for the
+	 * specified CCID.
+	 *
+	 * @param conn        Connection object used in
+	 *                    bt_cap_find_ccid_service().
+	 * @param err         Error value. 0 on success, GATT error on positive
+	 *                    value or errno on negative value.
+	 * @param ccid        The CCID value used in bt_cap_find_ccid_service().
+	 * @param ccid_record A CCID record structure with a reference to the
+	 *                    service instance. May be NULL, if not service
+	 *                    instance was found. Only valid if @p err is 0.
+	 */
+	void (*ccid_found)(struct bt_conn *conn, int err, uint8_t ccid,
+			   const struct bt_cap_ccid_record *ccid_record);
+}
+
+/**
+ * @brief Find the content control service based on CCID
+ *
+ * This is a local operation, and requires that the content control services
+ * already have been discovered using e.g. bt_tbs_client_discover() or
+ * media_proxy_ctrl_discover_player().
+ *
+ * @param conn Connection to a remote server or NULL to get local instance.
+ * @param ccid The CCID to find.
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_cap_find_ccid_service_service(struct bt_conn *conn, uint8_t ccid);
```

Comment:
Should't it be ccid_type instead? How the user knows ccid ?

Reply:
I am not sure I understand what you mean by `ccid_type`. 

The CCID value will be part of the metadata when setting up a stream as far as I can tell, and is provided the client setting up the stream. 

This is something that will need to be exposed with the unicast server API I suppose, but that part is still missing from that API. 

One thing I am also a bit unclear on, is how much of the API we should duplicate, e.g. if there's a call to a unicast server callback (pending change in the topic-le-audio branch), should the unicast server call a similar callback for the CAP acceptor? We currently do that for e.g. the ISO callbacks, which are currently duplicated by the audio stream implementation. 

I think we might want to add a metadata updated callback for the CAP acceptor, which would include the metadata itself, and thus the UUID, possibly deconstructed from the LTV list and put into a user-friendly struct instead. 

Let me add an update for that, and see if that makes sense.

Reply:
I'll wait with an (possible) update until we have resolved whether or not we want to duplicate the callbacks for CAP, when they already exists for BAP.

Reply:
Sorry for the shortcut, but maybe I misunderstood API here. 

When reading this PR, I see that it is planned that CAP is exposing context for TBS and MCS to the User. User can find the context for given bt_conn and ccid. Is that correct?

Indeed User on the Acceptor side might be interested in CCID for TBS and MCS and the context for those services, in order to do operations on that and also possible to have additional knowledge about who is controlling stream (as you pointed metadata in BAP).
So when looking into this, I thought that idea is that CAP will discover those services and make it easy to user to get those.
Having that in mind, I tried to understand how the user will get CCID to provide to `bt_cap_find_ccid_service_service`, as the only way is to actually read characteristic on TBS or MCS (depends which you are interested in). 

So I though, maybe it shall be  `bt_cap_ccid_record_type` instead?

Reply:
Sorry for late reply, must've missed your comments @rymanluk 

> When reading this PR, I see that it is planned that CAP is exposing context for TBS and MCS to the User. User can find the context for given bt_conn and ccid. Is that correct?

Correct, the find_by_ccid is a spec defined procedure for the acceptor to match the CCID from the BAP metadata to a MCS or TBS instance on the initiator. 

> So when looking into this, I thought that idea is that CAP will discover those services and make it easy to user to get those.

The CCID of those services are, as far as I know, not constant and may change. That being said, it is probably possible to have the acceptor (and commandar) implementation autonomously do this, which would make the acceptor 100% autonomous regarding CAP procedures (but not BAP procedures). 

> Having that in mind, I tried to understand how the user will get CCID to provide to bt_cap_find_ccid_service_service, as the only way is to actually read characteristic on TBS or MCS (depends which you are interested in).

The CCID will be in the metadata from BAP

> So I though, maybe it shall be bt_cap_ccid_record_type instead?

Not sure what you mean by that. Are you suggesting to rename the function? 

> Do we need to expose BAP to the User since we do expose CAP? I thought CAP will do all we need (using BAP)

As discussed in the meeting, we don't want to duplicate callbacks and functions (procedures) from other profile implementations, so to use CAP, the user will also need to register for the BAP callbacks.

Reply:
@Thalley we are good as agreed on the meeting, that idea is the CCID will be provided by metadata and then user will use it for the specific context.